### DISCUSSION TITLE
Duplicate connection issue when connecting in SwiftUI View

### DISCUSSION THREAD

**Discussion by rayriffy** [google/JacquardSDKiOS#8] [General]
I'm trying to create custom Jacquard app based on examples in this repository. I've to adapt code lifecycle a bit to make it to work with SiwftUI.

Expectation
---

Function called to [connect to the tag](https://github.com/rayriffy/jacquard-rtc/blob/main/Jacquard%20RTC/Views/ConnectingView.swift#L23-L35). Then [monitor connection status](https://github.com/rayriffy/jacquard-rtc/blob/726cce4c1db0e6602f156534bdadf67891c3e90f/Jacquard%20RTC/Views/ConnectingView.swift#L41-L59) before continue to the next view

Actual result
---

[During `.connect()`](https://github.com/rayriffy/jacquard-rtc/blob/main/Jacquard%20RTC/Views/ConnectingView.swift#L24) SDK encounter `JacquardSDK.TagConnectionError.internalError` exception before hit [`.catch` handler](https://github.com/rayriffy/jacquard-rtc/blob/main/Jacquard%20RTC/Views/ConnectingView.swift#L23-L35)

Repository of the project can be found [here](https://github.com/rayriffy/jacquard-rtc)

```
tag connect request
F8EBF235-DDE2-BAC5-E979-AE7CC1BFE210
nil
2022-08-05 23:49:58.415359+0700 Jacquard RTC[2345:545375] [UIFocus] _TtGC7SwiftUI14_UIHostingViewVS_7AnyView_ implements focusItemsInRect: - caching for linear focus movement is limited as long as this view is on screen.
[ERROR] No transition found for (bluetoothConnected, didConnect(<JacquardSDK.PeripheralImplementation: 0x282988940>))
[ERROR] No transition found for (disconnected, didDiscoverServices([<CBService: 0x283c20e40, isPrimary = YES, UUID = D2F2BF0D-D165-445C-B0E1-2D6B642EC57B>, <CBService: 0x283c20e80, isPrimary = YES, UUID = Battery>]))
[INFO] State machine is already terminal, ignoring event: didConnect(<JacquardSDK.PeripheralImplementation: 0x2829d4220>)
[ERROR] No transition found for (error(JacquardSDK.TagConnectionError.internalError), didConnect(<JacquardSDK.PeripheralImplementation: 0x2829d4220>))
```

**Comment by aufflick** [comment]
Hi, thanks for the question :)

A quick look at the state machine logs you pasted suggests that `connect` is being called more than once on the same instance.

```
[ERROR] No transition found for (bluetoothConnected, didConnect(<JacquardSDK.PeripheralImplementation: 0x282988940>))
```

What this is saying is that the current state of the tag state machine is `bluetoothConnected`, and yet it is being asked to handle the message `didConnect` -- which absent a bug in either the state machine or CoreBluetooth should only be possible if `connect` is called more than once.

You're calling`connect` in a SwiftUI View struct `init` - these are simple value objects which may be (re)created more than once so it's likely that's what's happening here.

In fact when I run your code connected to a single tag I see three instances of `tag connect request` in the console.

There are two ways you can solve this - one is to write a SwiftUI State Object wrapper for your connection/tag logic (ie. something that conforms to [ObservableObject](https://developer.apple.com/documentation/combine/observableobject). I've done this before and it works but feels like a bunch of boilerplate code.

Since you're targeting iOS 16 I'd suggest using the (new in iOS 15) `.task` SwiftUI modifier for calling async/await code for a view (SwiftUI connects very nicely with the new structured concurrency features). This means you'll need to convert the Combine publisher to an AsyncStream somehow. It's pretty easy to do that by hand (see below) but it should also be possible to write a generic bridging function.

Since you're using the tag ID connection method, something like this should work:

## Create an async/await wrapper for `connect`

```swift
extension JacquardManager {
  func asyncConnect(_ identifier: UUID) -> AsyncThrowingStream<TagConnectionState, Error> {
    AsyncThrowingStream { continuation in
      let cancellable = self.connect(identifier)
        .sink { completion in
          switch completion {
          case .finished:
            continuation.finish()
          case .failure(let error):
            continuation.finish(throwing: error)
          }
        } receiveValue: { connectionState in
          continuation.yield(connectionState)
        }
      
      continuation.onTermination = { @Sendable _ in
        cancellable.cancel()
      }
    }
  }
}
```

## Triggering connection and looping over the `AsyncThrowingSequence`

```swift
struct ConnectingView: View {
  @State private var connectedTag: ConnectedTag? = nil
  @State private var lastConnectionStateDescription = "Preparing to connect..."
  
  private var tagIdentifier: UUID
  private var jqManager: JacquardManager
  
  init(manager: JacquardManager, tagId: UUID) {
    self.tagIdentifier = tagId
    self.jqManager = manager
  }
  
  var body: some View {
    VStack {
      
      if let connectedTag = connectedTag {
        Text("Connected to \(connectedTag.displayName)")
      } else {
        Text(lastConnectionStateDescription)
      }
    }
    .task {
      do {
        let connectionStream = jqManager.asyncConnect(tagIdentifier)
        for try await connectionState in connectionStream {
          print("\(connectionState)")
          switch connectionState {
          case .connected(let tag):
            connectedTag = tag
          default:
            connectedTag = nil
            lastConnectionStateDescription = "\(connectionState)"
          }
        }
      } catch {
        print("error during connection: \(error)")
      }
    }
  }
}
```

Dropping this into your repository seems to work with an adhoc test, including tag reconnection (you'll want to retain and pass around the `connectionStream` instance not the `connectedTag` instance so that you handle reconnections like the Combine examples do. If you search the usual places you'll find some people have written helper functions for `AsyncSequence` that provide Combine/Rx like operators if that's helpful for your design.

**Comment by rayriffy** [comment]
I think your comment can now be marked as answer. I am now got everything running which is great. I'm just making some small app to send gesture to trigger keyboard press. [Here's a repo](https://github.com/rayriffy/jacquard-rtc)

Would be great if I see next major version of SDK to be modernize with async comes as a default
